### ISSUE TITLE
High CPU usage in Octoprint?

### ISSUE THREAD

**Issue by agentsmithitaly** [foosel/OctoPrint#328]
After printing for 10 hours with no flaws my Octoprint instance is suffering a memory leak / eating all CPU resources on a 256MB RPi.
Cannot find commit number however I did a git pull on the devel branch on 08/12/2013 @ 19.00. VERSION file contains version number 1.1.0-dev

Here the statistics by top:
![raspberry-memoryleak](https://f.cloud.github.com/assets/5151309/1751453/52b057ae-65e8-11e3-8feb-04415d87ba6d.jpg)

The instance is still responsive, i.e. it is still able to send manual commands to the printer and download timelapses. However system commands fail with the error: 
Command failed: OSError(12, 'Cannot allocate memory')

I noticed that the timelapse for the last print has not been generated, probably it is not a cause but an effect of the memory leak.

How can I help you in finding the cause?

There is this tool that graphs objects in Python but didn't really understood how to use it and its output
http://mg.pov.lt/objgraph/
Here's my config.yaml:

accessControl:
      enabled: false
    api:
      key: 6BB59A8C26364BA39EB1AAE0DA044470
    appearance: {}
    cura: {}
    feature:
      gCodeVisualizer: false
      sdSupport: false
    printerParameters:
      invertAxes:
      - z
      movementSpeed:
        x: 1000
        y: 1000
    serial:
      autoconnect: true
      port: /dev/ttyUSB0
      timeout:
        communication: 10.0
        connection: 10.0
    server:
      firstRun: false
      host: 127.0.0.1
    system:
      actions:
      - action: printer off
        command: printer.sh off
        confirm: You are about to turn off the printer.
        name: Turn off printer
      - action: printer on
        command: printer.sh on
        confirm: false
        name: Turn on printer
    temperature:
      profiles:
      - bed: '100'
        extruder: '195'
        name: ABS
      - bed: 60
        extruder: '160'
        name: PLA
    webcam:
      ffmpeg: /usr/bin/avconv
      snapshot: http://127.0.0.1:8080/?action=snapshot
      stream: /webcam/?action=stream
      timelapse:
        options:
          interval: 5
        type: timed

**Comment by foosel** [comment]
@agentsmithitaly Sorry for the long silence :/ I did try to reproduce it a couple of times, but I so far haven't manage to see anything like it. Do you still see it with the current version?

**Comment by foosel** [comment]
@agentsmithitaly ping?

**Comment by soswow** [comment]
I am using Raspberry PI 1 (2011.12) I am not sure when it started (maybe it was so from the beginning) I wasn't sure if it's my home network, bed position of wifi usb dongle or anything, but I started loosing connectivity with my RPi after some time after turning it on or starting a print.
I've decided to understand why it's so, and I've opened ssh. I see 97-100% CPU usage by octoprint right after service start. I don't have client open (browser).

At some point before I've disconnected camera module. I thought this was a problem. Seems not. I've set null for webcam: stream: because I saw something about CPU and streaming in FAQ. It didn't help.

Currently I do `apt-get update/upgrade`. Maybe this will help. 

At some point I've hit problem with my disk being full. I discovered it when I was trying to update octoprint. Maybe something related to this.

**Comment by nophead** [comment]
I have noticed OctoPrint does use very high CPU for a few minutes when it
starts up. It drops much lower after that. Not sure what it is doing but
the start up time seems much longer than it used to be.

**Comment by soswow** [comment]
The problem with connectivity dropping is definitely my network problems. I've concluded that off/on my computer wifi helps to re-establish connectivity.

I have a !feeling! that after upgrading apt-get not it doesn't take 100% when it does nothing.

**Comment by foosel** [comment]
When OctoPrint starts up it does some preemptive caching, because otherwise the first access by a new client right after the start takes long since the quite complex single page needs to be rendered first. During startup, that rendering is done and the results are cached. The reason being that your server probably starts up way before you try to access it, allowing you to get a faster response when you then in fact do access it.

Also, during start OctoPrint will look if there are any gcode files in the uploads folder it hasn't yet analysed and start analysis if there are any. Gcode analysis is also a quite cpu intensive process (which is why it's disabled while printing).

tldr: OctoPrint doses some cpu intensive stuff right on start up so that it doesn't need to do that later when the CPU resources might be needed for different tasks or the wait times might be annoying.

**Comment by nophead** [comment]
Problem is starting OctoPrint always coincides with wanting to print
something. Why would I boot up the Pi otherwise? Granted I might leave it
running for subsequent prints but I don't always have the forethought to
warm up OctoPrint in advance.

I have lots of gcode files but presumably they should all be analysed
already. Does it take a long time to decide if they are?

Fortunately the high CPU after the interface has loaded doesn't affect my
prints because my beds take longer to warm up. I noticed it when I was
using top to see how much CPU the RPI camera uses. The answer seems to be
lots unless you drop the framerate to 4 and reduce the quality and use a
wired connection instead of WiFi.

**Comment by foosel** [comment]
> Why would I boot up the Pi otherwise? Granted I might leave it running for subsequent prints

That's what most people do. It's a server application. Servers usually stay running. Considering the limited resources I have at my disposal on e.g. a Pi1, off-loading that stuff to server start vs. when you actually do access the page makes more sense for the majority of people out there (myself included btw). If you don't want it to preemptively cache since it doesn't suit your specific use case, you can always disable that feature. Set `devel.cache.preemptive` to `false` in `config.yaml`:

``` yaml
devel:
  cache:
    preemptive: false
```

> Does it take a long time to decide if they are?

No, but it needs to collect a list of all of your files to determine if there are any in there it hasn't seen yet, and I've seen that for whatever reason generating file listings on the Pi can be surprisingly CPU intensive (I don't know why and haven't looked into it further, it's not limited to python though, I've also seen that when doing name matching finds in bash). In any case, my guess would be that the large CPU spike you see is rather caused by the preemptive caching.

**Comment by nophead** [comment]
>Servers usually stay running.

Yes but when you have lots of them the electricity usage adds up. When I was churning out Mendel90 kits all my machines where busy but now they only get occasional use so I tend to turn them off.

**Comment by foosel** [comment]
Problem as originally reported by OP is still unreproduced, OP has also not responded again. High CPU usage during startup is normal as pointed out above (including how to disable preemptive caching). Closing this.
